// @ts-graphviz/common/lib/common.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
function _pipe(...args) {
  const [o1, ...operations] = args;
  return (...t) => operations.reduce((acc, f) => f(acc), o1(...t));
}
function pipe(o1, ...operations) {
  return _pipe(o1, ...operations);
}
const defer = (fn) => (...args) => (src) => fn(src, ...args);
const toIterable = (f) => ({
  [Symbol.iterator]: f
});
const map = defer(
  (src, selector) => Array.from(
    toIterable(function* () {
      for (const v of src) {
        yield selector(v);
      }
    })
  )
);
const filter = defer(
  (src, pred) => Array.from(
    toIterable(function* () {
      for (const x of src) {
        if (pred(x)) {
          yield x;
        }
      }
    })
  )
);
const RootModelsContext = Object.seal({
  // NOTE: RootModelsContext is also initialized after the model class is declared in the '@ts-graphviz/core/register-default' module.
  Graph: null,
  Digraph: null,
  Subgraph: null,
  Node: null,
  Edge: null
});
function createModelsContext(models) {
  return Object.assign(
    Object.seal(Object.assign({}, RootModelsContext)),
    models
  );
}
function isForwardRefNode(object) {
  return typeof object === "object" && object !== null && typeof object.id === "string";
}
function isNodeModel(object) {
  return typeof object === "object" && object !== null && object.$$type === "Node" && typeof object.id === "string";
}
function isNodeRef(node) {
  return isNodeModel(node) || isForwardRefNode(node);
}
function isNodeRefLike(node) {
  return typeof node === "string" || isNodeRef(node);
}
function isNodeRefGroupLike(target) {
  return Array.isArray(target) && target.every(isNodeRefLike);
}
function isCompass(c) {
  return ["n", "ne", "e", "se", "s", "sw", "w", "nw", "c"].includes(c);
}
function toNodeRef(target) {
  if (isNodeRef(target)) {
    return target;
  }
  const [id, port, compass] = target.split(":");
  if (isCompass(compass)) {
    return { id, port, compass };
  }
  return { id, port };
}
function toNodeRefGroup(targets) {
  if (targets.length < 2 && (isNodeRefLike(targets[0]) && isNodeRefLike(targets[1])) === false) {
    throw Error("EdgeTargets must have at least 2 elements.");
  }
  return targets.map((t) => toNodeRef(t));
}
exports.RootModelsContext = RootModelsContext;
exports.createModelsContext = createModelsContext;
exports.filter = filter;
exports.isCompass = isCompass;
exports.isForwardRefNode = isForwardRefNode;
exports.isNodeModel = isNodeModel;
exports.isNodeRef = isNodeRef;
exports.isNodeRefGroupLike = isNodeRefGroupLike;
exports.isNodeRefLike = isNodeRefLike;
exports.map = map;
exports.pipe = pipe;
exports.toNodeRef = toNodeRef;
exports.toNodeRefGroup = toNodeRefGroup;
